### PR TITLE
support lower_bound for iterator

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
@BusyJay @AndreMouche PTAL

### CODE REVIEW

File: src/coprocessor/dag/executor/scanner.rs [hunk 1]
```diff
@@ -75,15 +82,16 @@ impl Scanner {
...
-                self.statistics_cache.add(self.scanner.get_statistics());
-                let upper_bound = Some(Key::from_raw(&self.range.end).encoded().to_vec());
-                self.scanner = store.scanner(self.scan_mode, self.key_only, upper_bound)?;
-            }
+            ScanMode::Forward => self.seek_key = self.range.get_start().to_vec(),
             _ => unreachable!(),
         };
+
+        self.statistics_cache.add(self.scanner.get_statistics());
+        let lower_bound = Some(Key::from_raw(&self.range.start).encoded().to_vec());
```
Comment (AndreMouche) [inline src/coprocessor/dag/executor/scanner.rs]: It seems the same with L51-L65?

File: src/coprocessor/dag/executor/scanner.rs [hunk 2]
```diff
@@ -75,15 +82,16 @@ impl Scanner {
...
+        let upper_bound = Some(Key::from_raw(&self.range.end).encoded().to_vec());
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/scanner.rs]: What happen if it reverse seek `upper_bound`?

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 3]
```diff
@@ -156,24 +158,51 @@ pub struct RegionIterator {
...
+            } else {
+                region_start_key
+            }
+        }
+        _ => region_start_key,
+    };
+    iter_opt.set_lower_bound(lower_bound)
+}
+
 fn set_upper_bound(iter_opt: IterOption, region: &Region) -> IterOption {
```
Comment (BusyJay) [inline src/raftstore/coprocessor/region_snapshot.rs]: Can it take a `&mut IterOption` instead?

File: src/coprocessor/dag/executor/scanner.rs (hunk 2 above)
Comment (zhangjinpeng1987) [reply src/coprocessor/dag/executor/scanner.rs]: Inside the RocksDB, `SeekForPrev(upper_bound)` will return the largest valid key, it is equal to `SeekToLast()`. I see we use `SeekForPrev` to implement reverse seek, so the behavior is the same.

File: src/coprocessor/dag/executor/scanner.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply src/coprocessor/dag/executor/scanner.rs]: L51-L65 depends if `desc` is set, but here is not.

File: src/coprocessor/dag/executor/scanner.rs [hunk 4]
```diff
@@ -48,12 +48,18 @@ impl Scanner {
         key_only: bool,
         range: KeyRange,
     ) -> Result<Scanner> {
-        let (scan_mode, seek_key, upper_bound) = if desc {
-            (ScanMode::Backward, range.get_end().to_vec(), None)
+        let (scan_mode, seek_key, lower_bound, upper_bound) = if desc {
```
Comment (hicqu) [inline src/coprocessor/dag/executor/scanner.rs]: Here why not set both `lower_bound` and `upper_bound`, like we do at line 92?

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 5]
```diff
@@ -156,24 +158,51 @@ pub struct RegionIterator {
...
+}
+
+fn set_upper_bound(iter_opt: &mut IterOption, region: &Region) {
+    let region_end_key = keys::enc_end_key(region);
     let upper_bound = match iter_opt.upper_bound() {
-        Some(k) if !k.is_empty() => keys::data_key(k),
-        _ => keys::enc_end_key(region),
+        Some(k) if !k.is_empty() => {
+            let k = keys::data_key(k);
+            if k < region_end_key {
```
Comment (hicqu) [inline src/raftstore/coprocessor/region_snapshot.rs]: I think we can use `cmp::max`.

**Comment by hicqu** [comment]
LGTM.

**Comment by BusyJay** [comment]
LGTM

File: src/raftstore/coprocessor/region_snapshot.rs [hunk 6]
```diff
@@ -591,5 +616,22 @@ mod tests {
         let mut expect = test_data.clone();
         expect.reverse();
         assert_eq!(res, expect);
+
+        // test lower bound
+        let store = new_peer_storage(engine.clone(), raft_engine.clone(), &region);
```
Comment (AndreMouche) [inline src/raftstore/coprocessor/region_snapshot.rs]: Could we define an independent function to test lower bound?

**Pull request merged**